core/src/main/scala/kafka/log/Log.scala

```diff
@@ -609,24 +716,22 @@ class Log(val dir: File,
    * logSegment will be rolled if one of the following conditions met
    * <ol>
    * <li> The logSegment is full
-   * <li> The maxTime has elapsed
+   * <li> The maxTime has elapsed since the largest timestamp in the segment (or since the create time if there is no
+   * message having a timestamp)
    * <li> The index is full
    * </ol>
    * @return The currently active segment after (perhaps) rolling to a new segment
    */
   private def maybeRoll(messagesSize: Int): LogSegment = {
     val segment = activeSegment
+    val reachedRollMs = segment.timeWaitedForRoll(time.milliseconds) > config.segmentMs - segment.rollJitterMs
```

Comment:
Yes based on the timestamp of the first message sounds good. More precisely the time based log rolling will be based on the first message that has a timestamp if there is at least one such message. If there is no message that has a timestamp, the time based log rolling will be based on the create time, which is the same as previous behavior.